### PR TITLE
Collapsed fields collision in graphql

### LINKED ISSUE #13926: [graphQL] named part fields return nulls when Collapsed is checked
When you will add Named Part to the **ContentType** and also check "Collapse" checkbox, data is not returned using GraphQL.

Steps to reproduce:
1. create new **ContentPart** and add at least one field,
2. create new **ContentType** and add newly created named part, mark **Collapse** checkbox in part settings,
3. create new **ContentItem** and fill all fields,
4. enable **GraphQL** feature and try to query the data.

Here is the result:
image

I've just fixed this issue and will create PR soon.

Also there is an issue when you will try to add many parts with same field names, if those parts are also "collapsed" this will result with an exception:
image

To solve this problem, I've added a few methods to prevent fields name collisions:
- graphQL field name is build based on part it belongs to,
- graphQL field name is build using defined prefix,
- graphQL field name is build using defined suffix.

Here is simple example, first option is selected to prevent graphQL fields name colliosions:
[graphql.webm](https://github.com/OrchardCMS/OrchardCore/assets/6403130/397fd2bf-cfd0-49cf-839e-4ad00f679a15)

### PR DESCRIPTION
**Pull request by lampersky**
fixes #13926

### CODE REVIEW

**Comment by lampersky** [comment]
@sebastienros whenever you have time, please have a look into this PR, also here quick [video](https://github.com/OrchardCMS/OrchardCore/issues/13926#issuecomment-1716390317) was added to present those changes in PartSettings

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs [hunk 1]
```diff
@@ -221,5 +222,17 @@ public bool IsHiddenByDefault(ContentTypePartDefinition definition)
...
             return false;
         }
+
+        internal string GetFieldName(ContentTypePartDefinition definition, string partName, string fieldName)
+        {
+            var settings = definition.GetSettings<GraphQLContentTypePartSettings>();
+
+            if (settings.PreventFieldNameCollision)
+            {
+                return $"{partName.ToFieldName()}{fieldName.ToPascalCase()}";
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs:232]: No separators?

Comment (lampersky) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs:232]: @sebastienros yes, no separator, **partName.ToFieldName()** is producing **camelCase**, **fieldName.ToPascalCase()** is producing **PascalCase** when concatenated you will get some user friendly fields

example:
partName: MyCustomPart
fieldName: MyCustomField
will produce: **myCustomMyCustomField** graphQL field

image

**Comment by Piedone** [comment]
Do you want to get back to the review of this, @sebastienros?

**Comment by Piedone** [comment]
@sebastienros?

File: src/OrchardCore.Modules/OrchardCore.ContentTypes/GraphQL/Drivers/GraphQLContentTypePartSettingsDriver.cs [hunk 2]
```diff
@@ -1,35 +1,44 @@
...
 
 namespace OrchardCore.ContentTypes.GraphQL.Drivers
 {
     public class GraphQLContentTypePartSettingsDriver : ContentTypePartDefinitionDisplayDriver
     {
         private readonly GraphQLContentOptions _contentOptions;
+        private readonly IStringLocalizer S;
 
-        public GraphQLContentTypePartSettingsDriver(IOptions<GraphQLContentOptions> optionsAccessor)
+        public GraphQLContentTypePartSettingsDriver(IOptions<GraphQLContentOptions> optionsAccessor, IStringLocalizer<GraphQLContentTypePartSettingsDriver> stringLocalizer)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentTypes/GraphQL/Drivers/GraphQLContentTypePartSettingsDriver.cs:19]: I don't see it used.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Fields/ContentFieldsProvider.cs [hunk 3]
```diff
@@ -116,6 +116,11 @@ public FieldType GetField(ContentPartFieldDefinition field)
             };
         }
 
+        public bool HasField(ContentPartFieldDefinition field)
+        {
+            return _contentFieldTypeMappings.ContainsKey(field.FieldDefinition.Name);
+        }
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Fields/ContentFieldsProvider.cs:122]:
```suggestion
        public bool HasField(ContentPartFieldDefinition field) => _contentFieldTypeMappings.ContainsKey(field.FieldDefinition.Name);
```

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs (hunk 1 above)
Comment (hishamco) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs:232]:
```suggestion
                return partName.ToFieldName() + fieldName.ToPascalCase();
```

File: src/OrchardCore.Modules/OrchardCore.ContentTypes/Views/GraphQLContentTypePartSettings.Edit.cshtml [hunk 4]
```diff
@@ -11,29 +11,52 @@
...
+        const setFieldVisibility = (checked) => {
+            preventFieldNameCollisionCheckbox.parentElement.style.display = checked ? 'block' : 'none';
+            if (!checked) {
+                preventFieldNameCollisionCheckbox.checked = false;
+            }            
+        };
+        collapseCheckbox.addEventListener('change', (e) => {
+            setFieldVisibility(e.target.checked);
+        });
+        setFieldVisibility(@(collapsed.ToString().ToLower()));
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentTypes/Views/GraphQLContentTypePartSettings.Edit.cshtml:60]:
```suggestion
        const setFieldVisibility = (checked) => {
            preventFieldNameCollisionCheckbox.parentElement.style.display = checked ? 'block' : 'none';

            if (!checked) {
                preventFieldNameCollisionCheckbox.checked = false;
            }            
        };
        collapseCheckbox.addEventListener('change', (e) => setFieldVisibility(e.target.checked));

        setFieldVisibility(@(collapsed.ToString().ToLower()));
```

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Fields/ObjectGraphTypeFieldProvider.cs [hunk 5]
```diff
@@ -54,5 +50,20 @@ public FieldType GetField(ContentPartFieldDefinition field)
...
+            return _partObjectGraphTypes.GetOrAdd(field.FieldDefinition.Name,
+                partName => serviceProvider.GetService<IEnumerable<IObjectGraphType>>()?
+                    .FirstOrDefault(x => x.GetType().BaseType.GetGenericArguments().First().Name == partName)
+                );
+        }
+
+        public bool HasField(ContentPartFieldDefinition field)
+        {
+            return GetObjectGraphType(field) != null;
+        }
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Fields/ObjectGraphTypeFieldProvider.cs:67]:
```suggestion
        public bool HasField(ContentPartFieldDefinition field) =>GetObjectGraphType(field) != null;
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Review by hishamco — APPROVED** [review]

**Comment by hishamco** [comment]
@hyzx86 do you have any feedback on this PR?

**Comment by Piedone** [comment]
Please also merge soon before any conflicts can arise, if suitable.

**Comment by Piedone** [comment]
Can this be merged, then, Hisham?

**Comment by lampersky** [comment]
third merge conflict resolved, hurry up guys 😛

**Comment by Piedone** [comment]
So, I assume yes, this can be merged :D.

Sorry about that, and thank you!

**Pull request merged**